### ISSUE TITLE
Support non-Cargo projects

### ISSUE THREAD

**Issue by ojeda** [rust-lang/miri#1835]
Miri currently relies on Cargo and Xargo.

It would be nice to support non-Cargo projects, at least via documenting the expected flags, sysroot, etc.

**Comment by bjorn3** [comment]
I believe you need to compile everything (including the sysroot) with `-Zalways-encode-mir` and then for the executable that you want to test use `miri` in the place of the compiler. (There is no rustup shim for `miri` so you need to use `rustup which miri` to get it's path)

**Comment by bjorn3** [comment]
You can compile everything with `--emit metadata` to prevent unnecessary codegen that the default `--emit link` would do. Miri doesn't use the compiled object files anyway.

**Comment by RalfJung** [comment]
Non-cargo use will not be supported for the foreseeable future (we simply don't have the resources for that), but documentation PRs are always welcome and if there are concrete questions for the right invocations, I'm happy to help. :)

`cargo miri run -v` should also help; that will print the exact invocations of the Miri binary that our cargo wrapper generates.

**Comment by RalfJung** [comment]
Btw, the implementation of `./miri run` also invokes the driver directly, so you can use that for some inspiration as well.

**Comment by ojeda** [comment]
Thanks both of you! I will write some docs along the way then :)

**Comment by Yinwhe** [comment]
Does this issue have any further discussions? I also need to use `Miri` on non-cargo projects.

I think this is quite a useful feature, since Rust is integrated into many mix-language projects. And among these projects, Rust may be used in non-cargo workspaces.

**Comment by RalfJung** [comment]
Miri doesn't really work in mixed-language projects though, so it's unclear whether there's a good use-case here.

**Comment by ojeda** [comment]
One use case could be to run pure Rust tests in such projects. Another similar one is running Rust programs/tools/scripts/... that may not be part of the normal build of those projects (i.e. the "main" build may have FFI etc., but not other subprojects/parts).
